API re-grouping proposal

erictune opened the pull request:
tl;dr
- `Job` to move to `apis/batch/v1` for 1.2.
- `ScheduledJob` and `Workflow` to move to `apis/batch/v1` when they become `v1`.
- `DaemonSet` to move to `cluster/v1` when it graduates to v1.
- `podSecurityPolicy` to move into `cluster` prior to it becoming `v1`.
- no immediate plan to move other types out of their current locations.  prob wait for v2

erictune committed:
Proposal on api groups.

erictune committed:
Fix.

erictune commented:
@bgrant0607 @deads2k @soltysh @sdminonne 

I am sure this will cause some controversy, but want to get the discussion started so that `Job` has a chance to land in the right place for 1.2.

erictune commented:
@thockin

erictune commented:
@derekwaynecarr @smarterclayton @lavalamp

lavalamp commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+/apis/componentsStatus
+	componentStatus
+
+/apis/ingress
```
suggest "net"

smarterclayton replied:
I might have gone for "edge" or "boundary" or something that was less net specific because I think we will have actual network objects for policy eventually.

thockin replied:
what's wrong with /apis/net/{ingress, network, policy} ?

smarterclayton replied:
Just concern about stuttering.

On Tue, Jan 26, 2016 at 1:12 AM, Tim Hockin <notifications@github.com>
wrote:

> In docs/proposals/api-grouping.md
> <https://github.com/kubernetes/kubernetes/pull/20120#discussion_r50798516>
> :
>
> > +
> > +```
> > +/api/v1
> > +	persistentVolumeClaim
> > +	pod
> > +	podTemplate
> > +	service
> > +	serviceAccount
> > +	endpoints
> > +	event
> > +	secret
> > +
> > +/apis/componentsStatus
> > +	componentStatus
> > +
> > +/apis/ingress
>
> what's wrong with /apis/net/{ingress, network, policy} ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20120/files#r50798516>.
>

lavalamp commented on docs/proposals/api-grouping.md:
````diff
@@ -0,0 +1,375 @@
+Sketch of final groupings (not showing types that may remain for deprecation purposes):
+
+```
+/api/v1
````
Suggest not changing anything in the legacy "" group at this time.

lavalamp replied:
In particular, we can't remove e.g. "node" from v1. We can only do something like that across major version transitions. So it could move out in v2.

smarterclayton replied:
We *can* surface it in two groups at once though.

lavalamp replied:
Yes, that's true.

On Mon, Jan 25, 2016 at 3:51 PM, Clayton Coleman <notifications@github.com>
wrote:

> In docs/proposals/api-grouping.md
> <https://github.com/kubernetes/kubernetes/pull/20120#discussion_r50776135>
> :
>
> > +
> > +### Conclusion
> > +
> > +The "user vs admin" has a clear benefit for hosting providers in that they can chose to not expose
> > +a single group which contains a bunch of things they do not want to expose (a custom scheduler
> > +might get nodes information from a different API).
> > +
> > +The "Vertical" grouping seems like it fits well with "conways law" and that "vertical" is the pattern that
> > +people who build large extensions will tend to follow.
> > +
> > +Therefore, I suggest a combination of the "user vs admin" and the "vertical" groupings.
> > +
> > +Sketch of final groupings (not showing types that may remain for deprecation purposes):
> > +
> > +```
> > +/api/v1
>
> We *can* surface it in two groups at once though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20120/files#r50776135>.
>

erictune replied:
Agree we cannot remove it from v1.

But we could we:

1. install the same resource under a different group name, so it appears in two places
1. announce deprecation of the legacy group
1. after `x` months, `6 < x < 18` , grumble under our breath at anyone who still uses the legacy group.

erictune replied:
oh, clayton beat me to it.

smarterclayton replied:
Suggest you grumble menacingly under your breath, that may result in a
slightly higher uptick.

lavalamp commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+	persistentVolume
+	limitRange
+	resourceQuota
+	daemonSet
```
I think this belongs in long-running?

erictune replied:
this is contrasting those two options.

lavalamp commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+	scale
+	horizontalPodAutoscaler
+
+/apis/thirdParty
```
If anything belongs in "extensions" it's third-party object registration :)

lavalamp commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+- `ScheduledJob` and `Workflow` to move to `batch` when the become `v1`.
+- `DaemonSet` to move to `cluster/v1` when it graduates from v1.
+- `podSecurityPolicy` to move into `cluster` prior to it becoming `v1`.
+- no immediate plan to move other types out of their current locations.
```
+1

derekwaynecarr replied:
+1

smarterclayton commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+- `Pods` depend on `Secrets` for `imagePullSecrets`.
+- `ServiceAccounts` refer to `Secrets`.
+- `Pods` can depend on `ConfigMap` or `Secret` via env vars, etc.
+- `ReplicationControllers` depend on `Pods` via a `PodTemplate`.
```
HorizontalPodAutoscaler depends on the "Scale" subresource on objects in multiple groups - ReplicationController, Deployment, or PetSet

derekwaynecarr commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+	workflow
+	job
+
+/apis/longrunning/v1  # need better name
```
I struggled with naming on this too for quota discussion.  

Batch vs Transactional 
Batch vs Application
Batch vs Service

Are all splits that make sense to me, but not clear I like any of them.

thockin replied:
service?

job/service is a pretty obvious split, though a bit stuttery.

bgrant0607 replied:
+ @bprashanth @smarterclayton since we need to decide where to put PetSet

bgrant0607 replied:
We've also discussed "apps".

It occurred to me that another property common to APIs in this group is that they can be scaled (via scale subresource): scalableapps, scaledapps, scalable, elastic, elasticapps. I wouldn't call PetSet "elastic", though, since scaling may be slow/disruptive. That would distinguish them from DaemonSet, at least.

bgrant0607 replied:
Or "continuous", to distinguish from "batch".

bprashanth replied:
I just picked "apps" for the first cut of petset. I know what "batch" means because I'm used to hearing it, just like replicated, clustered etc, not sure that applies for "continuous"; I think it might be confusing without included context. We already have "autoscaling".

smarterclayton replied:
I like "service" or "continuous".  I'm lukewarm on apps (everything in every API groups could be part of an app, like jobs or scheduled jobs).  I think PetSets could very well be elastic - i.e. a cassandra can be rapidly changed, while a mysql is crusty.

thockin replied:
"serving" or "servers" ?

I can live with "service" or "services" or "apps".  We should be consistent
about pluralization here.  We already have "metrics" and "extensions".

On Tue, Apr 19, 2016 at 1:30 PM, Clayton Coleman <notifications@github.com>
wrote:

> In docs/proposals/api-grouping.md
> <https://github.com/kubernetes/kubernetes/pull/20120#discussion_r60305118>
> :
>
> > +  kubernetes provider certainly wants tight control over, and may want
> > +  to hide so that it can be implemented differently.
> > +- non-admins can focus their learning on the User group.
> > +
> > +### Vertical Grouping
> > +
> > +Group things that work with the same type of workloads.
> > +
> > +Sketch:
> > +```
> > +/apis/batch/v1
> > +	scheduledJob
> > +	workflow
> > +	job
> > +
> > +/apis/longrunning/v1  # need better name
>
> I like "service" or "continuous". I'm lukewarm on apps (everything in
> every API groups could be part of an app, like jobs or scheduled jobs). I
> think PetSets could very well be elastic - i.e. a cassandra can be rapidly
> changed, while a mysql is crusty.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20120/files/f8ad8db5d45876e5ca9570c4d1bc4d993d6bfccf#r60305118>
>

derekwaynecarr commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+
+
+- pro: we have three similar things.  let us group them.  yay!
+- con: what about `PetSet`?
```
Assume HPA would go here as well?

derekwaynecarr commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+/apis/longrunning/v1  # need better name
+	replicaSet
+	deployment
+	petSet
```
It's still not clear to me where "template/parameterizedList" goes in this breakdown.

I see pro and con to have it here or in cluster.

erictune replied:
I don't think it goes in cluster.  I want cluster to be this thing that you can hide entirely from a large portion of the user population, and they won't care.   I don't think that works for "template/parameterizedList".

bgrant0607 replied:
@derekwaynecarr Why would you put ParameterizedList in cluster?

derekwaynecarr replied:
I cannot recall what my rationale for `cluster` is which makes me think its not strong.  I think my main thought at the time was that things in `batch` and `longrunning` would potentially need to reference `ParameterizedList` so I was trying to reason about a potential group for things that are common to the two.

thockin commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+	job
+	replicaSet
+	scaleStatus
+	scale
```
dup

thockin commented on docs/proposals/api-grouping.md:
````diff
@@ -0,0 +1,375 @@
+
+Sketch:
+```
+/apis/batch/v1
````
I feel like "batch" has connotation of 2nd-classness.  What about apis/job/v1 ?  It's a little stuttery, but it's about different kinds of jobs.

erictune replied:
To me, batch is about a group of work items that are all done at the same time.  Various dictionary definitions support that. 

 I think that second-class connotation comes from Borg, but that most people will have that connotation.

Also `kubectl get batch/jobs` sounds great.

thockin commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+Rationale for `batch` group is that if you are running a batch job, you are likely to care about
+all three things (run-to-completion, start-at-a-time, and dependencies).
+
+Rationale for `DaemonSet` being in separate group is that typically only the cluster admin
```
I'm not sure I buy this, given the number of people who have asked how to achieve this result over the last year.

erictune replied:
I claim all of those people wanting daemonset are cluster admins and/or would be satisfied with Anti-Affinity when it is done, which will work with replicaSet.

bgrant0607 replied:
It is problematic to allow arbitrary users to create DaemonSets, except on dedicated machines, and the meaning of DaemonSet is inherently tied to Nodes, so it does make sense to separate it from the more purely user-oriented, application-oriented APIs.

thockin replied:
it is not immediately obvious why daemonset is problematic any more than
pods are

On Tue, Jan 26, 2016 at 4:55 PM, Brian Grant <notifications@github.com>
wrote:

> In docs/proposals/api-grouping.md
> <https://github.com/kubernetes/kubernetes/pull/20120#discussion_r50927286>
> :
>
> > +
> > +/apis/longrunning/v1  # need better name
> > +	replicaSet
> > +	deployment
> > +	petSet
> > +
> > +/apis/cluster/v1
> > +	daemonSet
> > +
> > +# ReplicationController is deprecated by ReplicaSet.
> > +```
> > +
> > +Rationale for `batch` group is that if you are running a batch job, you are likely to care about
> > +all three things (run-to-completion, start-at-a-time, and dependencies).
> > +
> > +Rationale for `DaemonSet` being in separate group is that typically only the cluster admin
>
> It is problematic to allow arbitrary users to create DaemonSets, except on
> dedicated machines, and the meaning of DaemonSet is inherently tied to
> Nodes, so it does make sense to separate it from the more purely
> user-oriented, application-oriented APIs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20120/files#r50927286>.
>

bgrant0607 commented:
Ref. previous proposed breakdown in #8190

bgrant0607 commented:
I suggest we focus on APIs that plan to graduate to v1 in 1.2, and related APIs that might influence what groups they land in.

That includes Job and HorizontalPodAutoscaler.

bgrant0607 commented on docs/proposals/api-grouping.md:
````diff
@@ -0,0 +1,375 @@
+```
+/api/v1
+/apis/extensions/v1beta1
+/apis/extensions/v1beta1
````
Why is extensions/v1beta1 listed twice?

erictune replied:
fixed.

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+/apis/abac/v0
+/apis/abac/v1beta1
+/apis/authorization/v1beta1
+/apis/componentconfig/v1alpha1/
```
componentconfig and other config schema families, like the scheduler config or kubeconfig, shouldn't be considered to be API groups.

erictune replied:
Removed.

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+	configMap
+
+/apis/extensions/v1beta1
+	scale
```
Scale is a subresource.

erictune replied:
removed

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+	endpoints
+	node
+	namespace
+	binding
```
Binding is a subresource.

erictune replied:
removed.

derekwaynecarr replied:
It shows as a top-level resource via API discovery.

The top-level resource only supports POST.

See https://github.com/kubernetes/kubernetes/blob/master/pkg/master/master.go#L331

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+	persistentVolume
+	persistentVolumeClaim
+	pod
+	podTemplate
```
We aren't using podTemplate yet, and likely won't in the near future.

bgrant0607 replied:
That is, as a top-level resource.

We share the type, embedded in all the controller types.

erictune replied:
Do you expect everything that currently has a podTemplate embedded in it will move to support a reference to a `template` object?

erictune replied:
Anyhow, removed.

bgrant0607 commented on docs/proposals/api-grouping.md:
````diff
@@ -0,0 +1,375 @@
+/apis/componentconfig/v1alpha1/
+```
+
+Those api groups have nearly 50 "top-level" types: types that are
````
Nits aside, yes, there are lots.

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+	daemonSet
+	job
+	ingress
+	clusterAutoscaler
```
Shouldn't exist.

erictune replied:
removed.

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+## Upcoming Additions
+
+New top level objects being considered:
+- `template` and `parameters` are added by #18215.  Other things do not
```
I'm hoping other types (e.g., PetSet) will depend on them.

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+- `Pods` with an `RBDVolumeSource` or `CephFSVolumeSource` or `FlexVolumeSource` depend on `Secrets` to store a secret.
+- `Pods` depend on `Secrets` for `imagePullSecrets`.
+- `ServiceAccounts` refer to `Secrets`.
+- `Pods` can depend on `ConfigMap` or `Secret` via env vars, etc.
```
One way to slice this is what the node (kubelet + kube-proxy) needs to understand. Those types are "core" in a way that others are not. 

* User execution concepts: Pod, Secret, ConfigMap, PersistentVolumeClaim
* Infrastructure concepts: Node, PersistentVolume
* Networking concepts: Service, Endpoints

erictune replied:
Added this text.

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+for any "third-party" APIs?
+
+Some reasons are:
+
```
Independent evolution / API versioning is the reason we finally created groups.

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+  ourselves honest, though.
+
+- The same reasons programs have packages: to enforce boundaries, and prevent coupling (taken from #635).
+- To allow api versioning at different rates (taken from #635)
```
This is the most important reason.

erictune replied:
Okay.  I called that out.  I need to think about what that implies for selecting the best grouping.

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+  - particularly for hosted implementations
+- To allow replacement of project-provided kinds with proprietary types
+  - PaaSes on top of Kubernetes might to do this.
+- So that users can focus on learning one group at a time.
```
This is probably second most important.

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+
+Some reasons are:
+
+- Groups will eventually allow extension by third parties that we as a core project
```
Third most important.

erictune committed:
Update.

erictune commented:
Added proposal from #8190

bgrant0607 commented:
Also ref #3806

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+	petSet
+	workflow
+	scheduledJob
+	componentStatus   # Maybe goes in apis/cluster?
```
Yes, would go in cluster

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+	workflow
+	scheduledJob
+	componentStatus   # Maybe goes in apis/cluster?
+	ingress           # Maybe goes in apis/cluster?
```
No, would not go in cluster

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+Rationale for `DaemonSet` being in separate group is that typically only the cluster admin
+needs to deal with DaemonSets.
+
+Other *verticals*, to use the Omega term, would be peers of these groupings.
```
Some discussion of verticals in #18880.

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+
+- pro: we have three similar things.  let us group them.  yay!
+- con: what about `PetSet`?
+- con: conflicts with *vertical* way of grouping things, and with *user vs admin*.
```
And with built-in vs. extension (e.g., DiurnalController)

erictune commented:
I addressed most of the comments.  I don't consider this complete though.

I want to reconsider the groupings in light of brian's suggestion of which reasons for grouping are most important.

Also, I didn't list any reasons to have fewer groups. 

Also, I'm worried about how verbose kubectl commands will become with group names (e.g. "kubectl get longrunning/replicasets").

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+
+### Respecting Dependencies
+
+If Object A1 in api group A strongly depends on Object B1 in api group B,
```
We will eventually need dependency management, not unlike packaging systems.

lavalamp commented:
We can give kubectl a group search order, so that if you omit the group
it's deterministic which one you'll get. We should optimize this so that
common resources find the right groups. ("pod" should find "/v1" not
"metrics/v1", for example).

On Tue, Jan 26, 2016 at 4:58 PM, Eric Tune <notifications@github.com> wrote:

> I addressed most of the comments. I don't consider this complete though.
>
> I want to reconsider the groupings in light of brian's suggestion of which
> reasons for grouping are most important.
>
> Also, I didn't list any reasons to have fewer groups.
>
> Also, I'm worried about how verbose kubectl commands will become with
> group names (e.g. "kubectl get longrunning/replicasets").
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20120#issuecomment-175318929>
> .
>

lavalamp commented:
(We should probably make people explicitly say the group when kubectl is
invoked from shell scripts, though.)

On Tue, Jan 26, 2016 at 5:30 PM, Daniel Smith <dbsmith@google.com> wrote:

> We can give kubectl a group search order, so that if you omit the group
> it's deterministic which one you'll get. We should optimize this so that
> common resources find the right groups. ("pod" should find "/v1" not
> "metrics/v1", for example).
>
> On Tue, Jan 26, 2016 at 4:58 PM, Eric Tune <notifications@github.com>
> wrote:
>
>> I addressed most of the comments. I don't consider this complete though.
>>
>> I want to reconsider the groupings in light of brian's suggestion of
>> which reasons for grouping are most important.
>>
>> Also, I didn't list any reasons to have fewer groups.
>>
>> Also, I'm worried about how verbose kubectl commands will become with
>> group names (e.g. "kubectl get longrunning/replicasets").
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/20120#issuecomment-175318929>
>> .
>>
>
>

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+and if api group A is "optional" then B should be "optional".
+But what does "strongly depend" mean?
+
+If Object C1 and Object C2 have a circular dependency, then it suggets that they belong in the same
```
Seems like a good rule.

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+/apis/cluster/v1
+	node
+	namespace
+	binding
```
The top-level binding path should go away, leaving only the binding subresources on pod and PVC.

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+	limitRange
+	resourceQuota
+	daemonSet
+	clusterAutoscaler
```
#17486

Also, don't see why it's listed twice

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+	clusterAutoscaler
+
+/apis/scaling/v1
+	scale
```
scale is a subresource that will be exposed by at least ReplicationController, ReplicaSet, and Deployment

bgrant0607 commented on docs/proposals/api-grouping.md:
````diff
@@ -0,0 +1,375 @@
+```
+
+
+Suggested order of work:
````
This proposal sounds fine.

bgrant0607 replied:
Though we need to decide what to do with HPA and maybe ThirdPartyResource.

bgrant0607 commented:
Reasons for fewer API groups:

More API groups is more complicated for users and client developers to reason about. 

Think about the extreme: We've discussed whether we should version every individual resource independently. That would provide maximum flexibility. But it's also hard to infer a coherent API surface. Every cross reference becomes a cross-group reference.

They are also more complex for K8s developers. They'd increase the test matrix -- think of all the version-skewed test combinations. And look at the issues with the scale subresource, which is of a different group/version than some of its parent resources.

For users, 2 groups would be the sweet spot: everything they care about, and everything they don't. If all users cared about the same things.

Really we want as few groups as possible -- just the ones that we think are necessary or important.

soltysh commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,427 @@
+Suggested order of work:
+
+- `Job` to move to `batch` for 1.2.
+- `ScheduledJob` and `Workflow` to move to `batch` when the become `v1`.
```
SGTM my only question is what or rather where newly creatd objects land. Since you're proposing group's layout how about adding a sentence about naming scheme for new object that will be written soon, ScheduledJob to name one. This should provide enough guidance in the topic of API structuring.

erictune committed:
Fix.

lavalamp commented:
@bgrant0607 

> They'd increase the test matrix -- think of all the version-skewed test combinations.

This is why we're generating client sets-- one version per group. It constrains the testing matrix. I imagine there will be a de-facto set on the server side, as well-- the default version for every group. We will *not* encourage people to make their servers into special snowflakes by incrementing default/storage versions one at a time; rather we'll make a recommendation with every release. Otherwise we'd have a testing combinatorial explosion even with very few group/versions. But with this rule I think more groups essentially isn't a test matrix problem.

erictune commented:
Filed new issue about name shortening and api groups: #20293

bgrant0607 commented:
The topic of categorization has reared its head again, this time for docs and UI: #22687

bgrant0607 commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,419 @@
+/apis/abac/v1beta1
+	policy // not currently available through API, but might be some day.
+
+/apis/authorization/v1beta1
```
Are these APIs or config?

deads2k replied:
> Are these APIs or config?

They are API.  They give information about the current state of a particular authorization decision.

smarterclayton commented on docs/proposals/api-grouping.md:
```diff
@@ -0,0 +1,375 @@
+	deployment
+	petSet
+
+/apis/cluster/v1
```
Would we ever have a cluster resource?  Would it go in "cluster" or "federation" API groups?
